src/Brick/Main.hs

```diff
@@ -316,6 +319,48 @@ findClickedExtents pos = EventM $ asks (findClickedExtents_ pos . latestExtents)
 findClickedExtents_ :: (Int, Int) -> [Extent n] -> [Extent n]
 findClickedExtents_ pos = reverse . filter (clickedExtent pos)
 
+-- | When processing a global EvMouseDown VtyEvent, the coordinates of
+-- the mouse event on the screen must be mapped to a specific location
+-- in a Widget.  The `lookupExtent` function will return the "extent"
+-- of the Widget (i.e. where it was drawn and how big it is) but this
+-- only indicates that the widget was clocked and does not identify
+-- the actual location within the widget where the click occurred.
+--
+-- The `getEventWidgetLocation` function translates the mouse event
+-- coordinates to a specific location within the widget in the
+-- widget's local reference frame, taking into account any scrolling
+-- that has occurred within a viewport that wraps that widget.
+--
+--    drawUI st = reportExtent (getName st) $
+--                viewport (getName st) Vertical $
+--                Widget Fixed Fixed $ ...
+--
+--    handleEvent event st =
+--      case event of
+--        EvMouseDown col row button _mods ->
+--          do wcoords <- getEventWidgetLocation st col row
+--             case wcoords of
+--               Nothing -> return st
+--               Just l -> ...
+--
+getEventWidgetLocation :: (Named a n, Ord n)
+                          => a -> Int -> Int -> EventM n (Maybe Location)
+getEventWidgetLocation widget screenCol screenRow =
+    do mExtent <- lookupExtent (getName widget)
+       case mExtent of
+         Nothing -> return Nothing
+         Just e@(Extent _ upperLeft _) ->
+             if clickedExtent (screenCol, screenRow) e
+             then let widgetRow = screenRow - upperLeft^.locationRowL
+                      widgetCol = screenCol - upperLeft^.locationColumnL
+                      widgetLoc = Location (widgetCol, widgetRow)
+                  in do mView <- lookupViewport (getName widget)
```

Comment:
I agree about the assumption of the naming, but the alternative is to explicitly pass the names which makes the interface more awkward.  My thinking was that if the user doesn't use the naming described in the documentation comment that they can use this function as a guide instead to implement a version more appropriate for their needs.

The use of this function is not limited to the Editor widget: in fact, I wrote it for use in my own Widget but thought it might be more generally applicable.